Display date and time in UTC, IST , PST

Surya-Santhosh opened the pull request:
Display date and time in UTC,IST,PST.
Write a C program to display the system’s current time in real-time. Time and date should be displayed in GMT, IST, PST timezone. Show Epoch time in GMT with date. Refresh the time every second. Create a main.c file to call these functions added in an other folder named appTimer. These functionality should be added in C and H files named appTimer.c and appTimer.h .Compile this file using GCC. Generate executable, object files, Assembly files, add proper gitignore. 

Update code to existing git repository in development branch using VScode UI

Surya-Santhosh committed:
Update main file to display date and time in UTC,IST,PST.
Add common.h file -Includes typdef.
Add appTimer folder -contain appTimer.c file and appTimer.h file .

Surya-Santhosh committed:
Update gitignore
Update main.c file - by removing windows.h and include unistd.h

Surya-Santhosh committed:
Correct the spelling of appTimer.h file

Surya-Santhosh committed:
change format specifier %02d to %02ld.
change prefix of variables

Surya-Santhosh committed:
Change the format specifier of ulEpoch to %ld

Surya-Santhosh committed:
declare variables into main.c file from global variables

Surya-Santhosh committed:
Remove windows.h file and use unistd.h

Surya-Santhosh committed:
Deleted already exist assembly and object files

adarshtrenser2023 commented on .gitignore:
```diff
@@ -1 +1,4 @@
+*.exe
+*.o
+*.s
+*.vscode
```
Should this .vscode needed here?

adarshtrenser2023 replied:
you can add  .vscode since if you stage it can be pushed, so better keep .vscode

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+#include<stdio.h>
+#include<stdbool.h>
+#include"appTimer.h"
+#include"../common.h"
```
stdbool and common are defined in appTimer.h please remove it.
Also use "common.h" without path

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+// Date    : 19/JUN/2024
+//
+//******************************************************************************
+//******************************* Include Files ********************************
```
new line

adarshtrenser2023 replied:
Please add new line between 12 and 13

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+//
+// File    : appTimer.c
+// Summary : Display date and time in UTC,IST,PST.
+// Note    : appTimer.h - macros, functions, global variables are included.
+// Note    : common.h - typdefs are included.
```
I dont think note should contain which all header file are included.

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+//***************************** Local Constants ********************************
+
+//***************************** Local Variables ********************************
+static uint8 ucFlag;
+static uint32 ulYear;
+static uint32 ulMonth;
+static uint32 ulHour;
+static uint32 ulMinute;
+static uint32 ulSeconds;
+static uint32 ulDay;
+static uint16 unIndex;
+static uint32 ulMonthName[MONTH] = {THIRTYONE_DAYS, TWENTYEIGHT_DAYS, 
```
what is the use of static keyword if the variable is global to the file.
also always initialize the variables

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+
+//******************************.appTimerUTC.***********************************
+// Purpose : Display date and time in UTC.
+// Inputs  : ulEpochUTC - Seconds in UTS.
```
there are two argument for this fucntion but here only 1 is mentioned

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+// Purpose : Display date and time in UTC.
+// Inputs  : ulEpochUTC - Seconds in UTS.
+// Outputs : None
+// Return  : true
```
the fucntion is also returning false at line 50 . Please add like "returns: false if ulEpochUTC is zero else True"

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+        return false;
+    }
+    ucFlag = ZERO;
+    ulYear = ulEpochUTC / (SECONDS * SECONDS * HOURS * DAYS);
```
macro name should be specific and what it does this is overall command
eg seconsd*seconds*hours*days , ithink seconds *minutes*hours*days is more correct

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+    }
+    ucFlag = ZERO;
+    ulYear = ulEpochUTC / (SECONDS * SECONDS * HOURS * DAYS);
+    ulEpochUTC = ulEpochUTC - (ulYear * (SECONDS * SECONDS * HOURS * DAYS));
```
instead of "a=a-b" use "a-=b"

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+
+    for (unIndex = UTC_YEAR; unIndex <= ulYear; unIndex++)
+    {
+        if ((unIndex % FOUR == ZERO && unIndex % HUNDRED != ZERO) || 
+        (unIndex % FOUR * HUNDRED) == ZERO)
```
1)using macro value is not just replacing numbers by string it should convey what the macro stands for:
what is four hundred 
also 2)as per standards condition check should be like zero ==
3) bracket for each condition is mandatory

adarshtrenser2023 replied:
Also indentation line 63 ( should be aligned to ( of 62

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+        if ((unIndex % FOUR == ZERO && unIndex % HUNDRED != ZERO) || 
+        (unIndex % FOUR * HUNDRED) == ZERO)
+        {
+            ucFlag++;
```
what is the use of ucFlag instead of incrementing here you can directly use ulDay-=1

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+        }
+    }
+    ulDay = ulDay - ucFlag;
+    for (unIndex = ZERO; unIndex < MONTH; unIndex++)
```
new line missing, check in complete file,

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+        if ((unIndex % FOUR == ZERO && unIndex % HUNDRED != ZERO) || 
+        (unIndex % FOUR * HUNDRED) == ZERO)
+        {
+            ulMonthName[ONE] = APRIL_DAYS;
```
April have only 29days?

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+    ulDay = ulDay - ucFlag;
+    for (unIndex = ZERO; unIndex < MONTH; unIndex++)
+    {
+        if ((unIndex % FOUR == ZERO && unIndex % HUNDRED != ZERO) || 
+        (unIndex % FOUR * HUNDRED) == ZERO)
```
this logic is invalid here since you are iterating only from 0 to 13 and checking for 4 , 400 100 %

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+        }
+        else
+        {
+            ulMonth = unIndex + ONE;
```
this can be updated outside the loop which can remove +one.

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+    printf("UTC (0:0)\n");
+    printf("-----------------------\n");
+    printf("Time  : %02ld:%02ld:%02ld ",ulHour,ulMinute,ulSeconds);
+    if (MONTH <= ulHour)
```
what is month here in logic , Work on all macros to have specific name for specific use.

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+// Return  : true
+// Notes   : None
+//****************************************************************************** 
+bool appTimerUTC(uint32 ulEpochUTC,uint32 ulEpoch)
```
what is the use of second argument

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+    printf("\nDate  : %02ld/%02ld/%02ld\n",ulDay,ulMonth,ulYear);
+    printf("Epoch  : %ld\n",ulEpoch);
+
+    return true;
```
every fucntion should have single return statement.

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,266 @@
+// Return  : true
+// Notes   : None
+//****************************************************************************** 
+bool appTimerIST(uint32 ulEpochIST)
```
Optimize the code with single function for converting all have same logic, 
try memory optimization.
remove unwanted variables remove unwanted macros.

adarshtrenser2023 commented on main.c:
```diff
@@ -1,18 +1,21 @@
-//******************************* Hello,World **********************************
+//******************************* TimerApp *************************************
```
file name is appTimer so keep name asAppTimer

adarshtrenser2023 commented on main.c:
```diff
@@ -23,16 +26,41 @@
+    uint32 ulEpoch;
+    uint32 ulEpochPST;
+    uint32 ulEpochIST;
+    uint32 ulEpochUTC;
```
initialze variables

adarshtrenser2023 commented on main.c:
```diff
@@ -23,16 +26,41 @@
+        struct tm *CurrentTime;
 
+        ulEpoch = time(&tTime);
+        CurrentTime = localtime(&tTime);
```
what is the use of CurrentTime here ? where is this used after geting localtime

adarshtrenser2023 commented on main.c:
```diff
@@ -23,16 +26,41 @@
+        ulEpochPST = ulEpoch - TIME_DIFF_PST;
+        appTimerPST(ulEpochPST);
+
+        sleep(THOUSAND);
```
is this 1000? Please show the output in next meeting

adarshtrenser2023 commented on main.c:
```diff
@@ -23,16 +26,41 @@
+
+        //PST
+        ulEpochPST = ulEpoch - TIME_DIFF_PST;
+        appTimerPST(ulEpochPST);
```
return is not used here for all the 3 APIs which can return true or false if not needed I think you should typecase with void , Not sure please check with Aromal

adarshtrenser2023 reviewed (commented):
Add comments  for better underatanding of logic

Surya-Santhosh committed:
Add main.c file

Surya-Santhosh committed:
Updated the code as per review.

Surya-Santhosh committed:
Changed prefix of time (time_t ulTimee)

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,167 @@
+    uint32 ulSeconds = ZERO;
+    uint32 ulDay = ZERO;
+    uint16 unIndex = ZERO;
+    bool blResult;
```
initialize with 0 no macro needed for that

adarshtrenser2023 replied:
blResult  not initialized

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,167 @@
+                ulDay--;
+            }
+        }
+        ulDay += ONE;
```
what is the use of this ? any specific reason
no comments in code , already asked for comments

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,167 @@
+        }
+
+        printf("\nDate  : %02ld/%02ld/%02ld\n",ulDay,ulMonth,ulYear);
+        blResult = true;
```
new line needed

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,167 @@
+                } 
+            }
+        }
+        ulDay += ONE;
```
new line after }

Surya-Santhosh committed:
Updated the code as per the review comments.
Initialize variables with 0.
add vscode in gitignore
add comments.

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,165 @@
+    {
+        ulYear = ulEpoch / (SECONDS * MINUTES * HOURS * DAYS);
+        ulEpoch -= (ulYear * (SECONDS * MINUTES * HOURS * DAYS));
+        ulYear += UTC_YEAR;
```
this function is generic so it should not contain any varibale or macro specific to UTC
also  what is the meaning of UTC_YEAR?

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,165 @@
+        {
+            if (HALFYEAR_MONTH <= unIndex)
+            {
+                if ((unIndex % EVEN) == 0)
```
keep standards and be consistant someplace  using 0 some place using ZERO

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,165 @@
+                    }
+                    else
+                    {
+                        ulMonth = unIndex;
```
this can be updated outside the loop , I think

adarshtrenser2023 commented on appTimer/appTimer.h:
```diff
@@ -0,0 +1,51 @@
+#define EVEN                       (2)
+#define ONE                        (1)
+#define ZERO                       (0)
+#define LEAPYEAR_CHECK_FOUR        (4)
```
Already mentioned macro naming should be specific, ONE, ZERO, EVEN
second _month value is 1 why it is then second month, it should  be 2 as per name

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,165 @@
+        {
+            if (((ZERO == (unIndex % LEAPYEAR_CHECK_FOUR)) && 
+            (ZERO != (unIndex % LEAPYEAR_CHECK_HUNDRED))) || 
+            (ZERO == (unIndex % LEAPYEAR_CHECK_FOUR * LEAPYEAR_CHECK_HUNDRED)))
```
i think there need to be a macro for checking 400, it is not good practice to use mutliple macro arithmetic's to attain a value code will work this is for readability from the name the logic need to be conveyed that is how you should update macro names

adarshtrenser2023 commented on appTimer/appTimer.h:
```diff
@@ -0,0 +1,51 @@
+#define SECONDS                    (60)
+#define MINUTES                    (60)
+#define HOURS                      (24)
+#define DAYS                       (365)
```
DAYS be specific  TOTAL_DAYS_IN_YEARS like that

adarshtrenser2023 reviewed (commented):
Consider replying to all the comments
Also do a self review , seen already mentioned issues are repeating. avoid that.
Do a full check to close all review comments by next commit

Surya-Santhosh committed:
Update the code as per review comments
1. Renamed macros.
2. Add new line. Before while.
3. Added comments.
4. Added check condition for function.

Surya-Santhosh committed:
Delete the txt files and added .txt in gitignore

adarshtrenser2023 commented on main.c:
```diff
@@ -23,16 +27,68 @@
+        {
+            printf("Epoch Time is Zero.");
+        }
+        printf("Epoch  : %ld\n",ulEpoch);
```
new line

Surya-Santhosh committed:
Added new line after if loop

Surya-Santhosh committed:
aligned tabs

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2025 Trenser Technology Solutions
+// All Rights Reserved
+//******************************************************************************
+//
```
Please check if there is new line needed at line 5 and 11

adarshtrenser2023 commented on appTimer/appTimer.c:
```diff
@@ -0,0 +1,163 @@
+//
+// File    : appTimer.c
+// Summary : Display date and time in UTC,IST,PST.
+// Note    : appTimer.h, common.h are included.
```
no need to add headers in notes, if nothing to be specified in notes you can put None.

Surya-Santhosh committed:
Remove extra lines in Function header block in appTimer.c,main.c
Remove Header files in note.

adarshtrenser2023 commented on main.c:
```diff
@@ -23,16 +25,69 @@
+        printf("UTC (0:0)\n");
+        printf("-----------------------\n");
+
+        if (appTimer(ulEpochUTC) == false)
```
false should start at left

Surya-Santhosh committed:
Updated the code as per review coments.
Chnaged if condition.

Surya-Santhosh committed:
Change the format specifier of unsigned long fron %ld to %lu.

Surya-Santhosh committed:
Updated the code as per review.
Changed if condition.
